src/main/sensors/compass.c

```diff
@@ -388,31 +404,245 @@ uint32_t compassUpdate(timeUs_t currentTimeUs)
+ *  P(2,2) = P(2,2) - g(2) * dP(2);
+ *  P(3,1) = P(3,1) - g(3) * dP(1);
+ *  P(3,2) = P(3,2) - g(3) * dP(2);
+ *  P(3,3) = P(3,3) - g(3) * dP(3);
+ *
+ *  // iteration 3
+ *  dP = [P(1,1)*gyro(i,2) - P(2,1)*gyro(i,1); ...
+ *        P(2,1)*gyro(i,2) - P(2,2)*gyro(i,1); ...
+ *        P(3,1)*gyro(i,2) - P(3,2)*gyro(i,1)];
+ *  zn(3) = 1/(lambda + gyro(i,2)*dP(1) - gyro(i,1)*dP(2));
+ *  g(1) = zn(3) * dP(1);
+ *  g(2) = zn(3) * dP(2);
+ *  g(3) = zn(3) * dP(3);
+ *  b(1) = b(1) - e(3) * g(1);
+ *  b(2) = b(2) - e(3) * g(2);
+ *  b(3) = b(3) - e(3) * g(3);
+ *  P(1,1) = P(1,1) - g(1) * dP(1);
+ *  P(2,1) = P(2,1) - g(2) * dP(1);
+ *  P(2,2) = P(2,2) - g(2) * dP(2);
+ *  P(3,1) = P(3,1) - g(3) * dP(1);
+ *  P(3,2) = P(3,2) - g(3) * dP(2);
+ *  P(3,3) = P(3,3) - g(3) * dP(3);
+
+ *  zn = zn * lambda;
+ *  P = P / lambda;
+ *  lambda = lambda_min + (1 - lambda_min) * ( zn.' * zn ) / 3.0;
+ */
+
+void compassBiasEstimatorInit(compassBiasEstimator_t *cBE, const float lambda_min, const float p0)
+{
+    compassBiasEstimatorUpdate(cBE, lambda_min, p0);
+    
+    cBE->lambda = lambda_min;
+
+    // don't know if it is necessary to initialize these as 0
+    for (unsigned i = 0; i < 3; i++) {
+        cBE->b[i] = 0.0f;
+        for (unsigned j = 0; j < 3; j++) {
+            if (i != j) {
+                cBE->P[i][j] = 0.0f;
+            }
+        }
+    }  
+}
+
+void compassBiasEstimatorUpdate(compassBiasEstimator_t *cBE, const float lambda_min, const float p0)
+{
+    cBE->lambda_min = lambda_min;
+    cBE->p0 = p0;
+
+    // update diagonal entries for faster convergence
+    for (unsigned i = 0; i < 3; i++) {
+        cBE->P[i][i] = cBE->p0;
+    } 
+}
+
+void compassBiasEstimatorApply(compassBiasEstimator_t *cBE, float *mag, const float *dmag, const float *gyro)
+{
+    //  e = dmag + cross(gyro, mag - b)
+    const float e[3] = {dmag[0] + gyro[2] * (cBE->b[1] - mag[1]) - gyro[1] * (cBE->b[2] - mag[2]),
+                        dmag[1] - gyro[2] * (cBE->b[0] - mag[0]) + gyro[0] * (cBE->b[2] - mag[2]),
+                        dmag[2] + gyro[1] * (cBE->b[0] - mag[0]) - gyro[0] * (cBE->b[1] - mag[1])};
+
+    float zn[3];
+
+    // iteration 1: k = 0; i = 1; j = 2; sign =  1.0f;
+    compassBiasEstimatorSolveIterative(cBE, e, zn, gyro, 0, 1, 2,  1.0f);
+    // iteration 2: k = 1; i = 0; j = 2; sign = -1.0f;
+    compassBiasEstimatorSolveIterative(cBE, e, zn, gyro, 1, 0, 2, -1.0f);
+    // iteration 3: k = 2; i = 0; j = 1; sign =  1.0f;
+    compassBiasEstimatorSolveIterative(cBE, e, zn, gyro, 2, 0, 1,  1.0f);
+
+    // zn = zn * lambda
+    // P = P / lambda
+    for (unsigned i = 0; i < 3; i++) {
+        zn[i] *= cBE->lambda;
+        for (unsigned j = 0; j < 3; j++) {
+            cBE->P[i][j] /= cBE->lambda;
+        }
+    }
+
+    // lambda = lambda_min + (1 - lambda_min) * ( zn.' * zn ) / 3.0
+    cBE->lambda = cBE->lambda_min - (cBE->lambda_min - 1.0f) * (zn[0] * zn[0] + zn[1] * zn[1] + zn[2] * zn[2]) / 3.0f;
```

Comment:
```suggestion
    cBE->lambda = cBE->lambda_min + (1.0f - cBE->lambda_min) * (zn[0] * zn[0] + zn[1] * zn[1] + zn[2] * zn[2]) / 3.0f;
```
Or change comment to match code

Reply:
I updated the comment, thx.